ocaml/gencert/selfcert.ml

```diff
@@ -0,0 +1,125 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module C = Cmdliner
+module Rsa = Mirage_crypto_pk.Rsa
+module UX = Xapi_stdext_unix.Unixext
+open Rresult (* introduces >>= >>| and R *)
+
+(** initialize the random number generator at program startup when this
+module is loaded. *)
+let () = Mirage_crypto_rng_unix.initialize ()
+
+let defer f = Fun.protect ~finally:f
+
+(** [write_cert] writes a PEM file to [path]. It attempts to do that
+  atomically by writing to a temporary file in the same directory first and
+  renaming the file at the end *)
+let write_certs path key cert =
+  let delim = "\n" in
+  let temp_dir = Filename.dirname path in
+  let tmp = Filename.temp_file ~temp_dir "certify-" ".tmp" in
+  try
+    let fd = Unix.openfile tmp [Unix.O_WRONLY] 0o600 in
+    defer (fun () -> Unix.close fd) @@ fun () ->
+    UX.really_write fd (Cstruct.to_string key) 0 (Cstruct.len key) ;
+    UX.really_write_string fd delim ;
+    UX.really_write fd (Cstruct.to_string cert) 0 (Cstruct.len cert) ;
+    Unix.rename tmp path ;
+    R.ok ()
+  with e -> R.error_msg (Printexc.to_string e)
+
+let expire_in days =
+  let seconds = days * 24 * 60 * 60 in
+  let start = Ptime_clock.now () in
+  match Ptime.(add_span start @@ Span.of_int_s seconds) with
+  | Some expire ->
+      R.ok (start, expire)
+  | None ->
+      R.error_msgf "can't represent %d as time span" days
+
+let add_dns_names extension = function
+  | [] ->
+      extension
+  | names ->
+      X509.Extension.(
+        add Subject_alt_name
+          (false, X509.General_name.(singleton DNS names))
+          extension)
+
+let sign days key pubkey issuer req alt_names =
+  expire_in days >>= fun (valid_from, valid_until) ->
+  match (key, pubkey) with
+  | `RSA priv, `RSA pub when Rsa.pub_of_priv priv = pub ->
+      let extensions = add_dns_names X509.Extension.empty alt_names in
+      X509.Signing_request.sign ~valid_from ~valid_until ~extensions req key
+        issuer
+      |> R.reword_error (fun _ -> Printf.sprintf "signing failed" |> R.msg)
+  | _ ->
+      R.error_msgf "public/private keys don't match (%s)" __LOC__
+
+let selfsign name alt_names length days certfile =
+  let rsa = Rsa.generate ~bits:length () in
+  let privkey = `RSA rsa in
+  let pubkey = `RSA (Rsa.pub_of_priv rsa) in
+  let issuer =
+    [X509.Distinguished_name.(Relative_distinguished_name.singleton (CN name))]
+  in
+  let req = X509.Signing_request.create issuer privkey in
+  sign days privkey pubkey issuer req alt_names >>= fun cert ->
+  let cert_pem = X509.Certificate.encode_pem cert in
+  let key_pem = X509.Private_key.encode_pem privkey in
+  write_certs certfile key_pem cert_pem
+
+let host name alt_names pemfile =
+  let expire_days = 3650 in
```

Comment:
I assume the 10 years here is meant to match our LTSR lifetime, so if your certs expire you should at least upgrade by that time, right?

What happens if someone's clocks go wrong though on boot (e.g. failing NVRAM battery, and no internet connection to synchronize the clock). Would that prevent the pool members from communicating with each-other (we do recommend using NTP, which should mostly solve this issue). I think we have to be sure to run the certificate generation *after* we got a stable time from NTP though.

Reply:
This is what xapi does so far:
```
ocaml/gencert/selfcert.ml:  let expire_days = 3650 in
```